test/monitoring/control-plane_test.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build packet
+// +build prometheus
+
+package monitoring
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	testutil "github.com/kinvolk/lokomotive/test/components/util"
+
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+func portForward(t *testing.T, readyChan, stopChan chan struct{}, closeChan func()) {
+	defer closeChan()
+
+	kubeconfig := testutil.KubeconfigPath(t)
+	t.Logf("using KUBECONFIG=%s", kubeconfig)
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		t.Fatalf("could not build config from KUBECONFIG: %v", err)
+	}
+
+	roundTripper, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		t.Fatalf("could not create round tripper: %v", err)
+	}
+
+	// Port forwarding only works with a pod. Even if we ask user to provide the service name, we
+	// still have to figure out the pod name and select one of the pod the service is pointing at.
+	// Instead we can rely on the fact that prometheus pods are started by statefulset and will
+	// always have consistent naming.
+	podName := "prometheus-prometheus-operator-prometheus-0"
+	namespace := "monitoring"
+
+	serverURL, err := url.Parse(config.Host)
+	if err != nil {
+		t.Fatalf("could not parse the URL from kubeconfig: %v", err)
+	}
+
+	serverURL.Path = fmt.Sprintf("/api/v1/namespaces/%s/pods/%s/portforward", namespace, podName)
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, serverURL)
+
+	// TODO: randomly select a port
+	ports := []string{"9090:9090"}
```

Comment:
If you use `0:9090` it should forward on random port. Then you can use https://godoc.org/k8s.io/client-go/tools/portforward#PortForwarder.GetPorts to get forwarded port.

Reply:
done